source/concurrent/WaitLock.ooc

```diff
@@ -0,0 +1,53 @@
+/* This file is part of magic-sdk, an sdk for the open source programming language magic.
+ *
+ * Copyright (C) 2016 magic-lang
+ *
+ * This software may be modified and distributed under the terms
+ * of the MIT license.  See the LICENSE file for details.
+ */
+
+use concurrent
+
+WaitLock: class {
+	_waitCondition := WaitCondition new()
+	_mutex : Mutex
+	_ownsMutex : Bool
+	init: func {
+		this _mutex = Mutex new()
+		this _ownsMutex = true
+	}
+	init: func ~WithMutex (mutex: Mutex)
+	{
+		this _mutex = mutex
+		this _ownsMutex = false
+	}
```

Comment:
You could have `_ownsMutex` to default to `false` and do something like
```ooc
init: func (mutex: Mutex = null) {
    if (mutex)
        this _mutex = mutex
    else {
        this _mutex = Mutex new()
        this _ownsMutex = true
    }
}
```
to avoid the need for two different constructors